feat(txpool): add sub-pools length and size metrics

leovct opened the pull request:
## Description

Add pool and sub-pools (`pending`, `basefee` and `queued`) length and size metrics.
Closes #2592 

## Test

- TODO

leovct commented on crates/transaction-pool/src/pool/txpool.rs:
```diff
@@ -253,9 +253,74 @@ impl<T: TransactionOrdering> TxPool<T> {
+        match total_number_transactions.try_into() {
+            Ok(val) => self.metrics.total_number_transactions.absolute(val),
+            Err(_) => {
+                tracing::debug!(target: "transaction-pool::pool", "Number of transactions in the pool too large to convert to u64")
```
Not sure about this, i'm quite new to rust

onbjerg commented on crates/transaction-pool/src/metrics.rs:
```diff
@@ -13,4 +13,23 @@ pub struct TxPoolMetrics {
     pub(crate) invalid_transactions: Counter,
     /// Number of removed transactions from the pool
     pub(crate) removed_transactions: Counter,
+    /// Total number of transactions in the pool
+    pub(crate) total_number_transactions: Gauge,
```
I think this can be inferred using `inserted_transactions - removed_transactions - invalid_transactions`, so no need for a separate metric.

Alternatively, it can be inferred using `pending_pool_length + basefee_pool_length + queued_pool_length` etc.

leovct replied:
Thanks for the review @onbjerg, it's very instructive! :)
I renamed the metrics, removed the ones that can be inferred and updated the documentation.

onbjerg commented on crates/transaction-pool/src/metrics.rs:
```diff
@@ -13,4 +13,23 @@ pub struct TxPoolMetrics {
+    /// Total amount of memory used by the transactions in the pool in bytes
+    pub(crate) total_size_bytes: Gauge,
+
+    /// Number of transactions in the pending sub-pool
+    pub(crate) pending_pool_length: Gauge,
```
```suggestion
    /// Number of transactions in the pending sub-pool
    pub(crate) pending_pool_transactions: Gauge,
```

onbjerg commented on crates/transaction-pool/src/metrics.rs:
```diff
@@ -13,4 +13,23 @@ pub struct TxPoolMetrics {
+    /// Number of transactions in the pending sub-pool
+    pub(crate) pending_pool_length: Gauge,
+    /// Total amount of memory used by the transactions in the pending sub-pool in bytes
+    pub(crate) pending_sub_pool_size_bytes: Gauge,
```
```suggestion
    pub(crate) pending_pool_size_bytes: Gauge,
```

onbjerg commented on crates/transaction-pool/src/metrics.rs:
```diff
@@ -13,4 +13,23 @@ pub struct TxPoolMetrics {
+    pub(crate) pending_sub_pool_size_bytes: Gauge,
+
+    /// Number of transactions in the basefee sub-pool
+    pub(crate) basefee_pool_length: Gauge,
```
```suggestion
    pub(crate) basefee_pool_transactions: Gauge,
```

onbjerg commented on crates/transaction-pool/src/metrics.rs:
```diff
@@ -13,4 +13,23 @@ pub struct TxPoolMetrics {
+    /// Number of transactions in the basefee sub-pool
+    pub(crate) basefee_pool_length: Gauge,
+    /// Total amount of memory used by the transactions in the basefee sub-pool in bytes
+    pub(crate) basefee_sub_pool_size_bytes: Gauge,
```
```suggestion
    pub(crate) basefee_pool_size_bytes: Gauge,
```

onbjerg commented on crates/transaction-pool/src/metrics.rs:
```diff
@@ -13,4 +13,23 @@ pub struct TxPoolMetrics {
+    pub(crate) basefee_sub_pool_size_bytes: Gauge,
+
+    /// Number of transactions in the queued sub-pool
+    pub(crate) queued_pool_length: Gauge,
```
```suggestion
    pub(crate) queued_pool_transactions: Gauge,
```

onbjerg commented on crates/transaction-pool/src/metrics.rs:
```diff
@@ -13,4 +13,23 @@ pub struct TxPoolMetrics {
+    /// Number of transactions in the queued sub-pool
+    pub(crate) queued_pool_length: Gauge,
+    /// Total amount of memory used by the transactions in the queued sub-pool in bytes
+    pub(crate) queued_sub_pool_size_bytes: Gauge,
```
```suggestion
    pub(crate) queued_pool_size_bytes: Gauge,
```

onbjerg commented on crates/transaction-pool/src/pool/txpool.rs:
```diff
@@ -253,6 +257,17 @@ impl<T: TransactionOrdering> TxPool<T> {
         let UpdateOutcome { promoted, discarded } = self.process_updates(updates);
 
+        // Compute pool metrics
+        let stats = self.size();
+        self.metrics.total_number_transactions.set(stats.total_transactions as f64);
+        self.metrics.total_size_bytes.set(stats.total_size as f64);
+        self.metrics.pending_pool_length.set(stats.pending as f64);
+        self.metrics.pending_sub_pool_size_bytes.set(stats.pending_size as f64);
+        self.metrics.basefee_pool_length.set(stats.basefee as f64);
+        self.metrics.basefee_sub_pool_size_bytes.set(stats.basefee_size as f64);
+        self.metrics.queued_pool_length.set(stats.queued as f64);
+        self.metrics.queued_sub_pool_size_bytes.set(stats.queued_size as f64);
```
I guess this is ok cc @mattsse

mattsse replied:
this seems ok!
looks correct

onbjerg reviewed (changes requested):
Some of the metrics are redundant since Prometheus is more than capable of adding up different series to infer these metrics, so we should eliminate them.

My other comments are mostly around naming; the names themselves don't matter too much but we should avoid terms like "length" and we should be somewhat consistent in the naming of the metrics

Thanks for taking this on!

leovct commented:
> Some of the metrics are redundant since Prometheus is more than capable of adding up different series to infer these metrics, so we should eliminate them.
> 
> My other comments are mostly around naming; the names themselves don't matter too much but we should avoid terms like "length" and we should be somewhat consistent in the naming of the metrics
> 
> Thanks for taking this on!

Thanks for the review @onbjerg, it's very instructive! :)
I renamed the metrics, removed the ones that can be inferred and updated the documentation.

mattsse commented on crates/transaction-pool/src/pool/txpool.rs:
```diff
@@ -253,6 +253,15 @@ impl<T: TransactionOrdering> TxPool<T> {
         // Process the sub-pool updates
         let UpdateOutcome { promoted, discarded } = self.process_updates(updates);
 
+        // Compute pool metrics
+        let stats = self.size();
+        self.metrics.pending_pool_transactions.set(stats.pending as f64);
+        self.metrics.pending_pool_size_bytes.set(stats.pending_size as f64);
+        self.metrics.basefee_pool_transactions.set(stats.basefee as f64);
+        self.metrics.basefee_pool_size_bytes.set(stats.basefee_size as f64);
+        self.metrics.queued_pool_transactions.set(stats.queued as f64);
+        self.metrics.queued_pool_size_bytes.set(stats.queued_size as f64);
+
```
this is a pedantic nit, but I'd like to do this in a separate function, like `update_size_metrics` or smth.

leovct replied:
done :)

mattsse reviewed (approved):
lgtm
